src/ray/object_manager/plasma/client.cc

```diff
@@ -297,7 +297,7 @@ void PlasmaClient::Impl::InsertObjectInUse(const ObjectID &object_id,
 
   // Add this object ID to the hash table of object IDs in use. The
   // corresponding call to free happens in PlasmaClient::Release.
-  it->second->object = *object.release();
+  it->second->object = std::move(*object);
```

Comment:
Can't this cause a double free depending on the implementation of the `PlasmaObject` move constructor? Because the `it->second->object` can contain a reference to the memory that is still owned by the `unique_ptr` (which will be freed after this function exits). Might be missing something, not a cpp expert.

Reply:
I think this is currently ok because the `PlasmaObject` is a struct only containing ints/bools, but would be better to explicitly copy the `PlasmaObject` instead

Reply:
`release` gets the pointer, `move` calls the move constructor here but (1) the object pointed by the pointer and (2) the pointer itself is still valid

Reply:
Just imagine you have a noop move constructor and assignment, do you think it's ok to delete the pointer here?
It's always yes.

Reply:
> I think this is currently ok because the PlasmaObject is a struct only containing ints/bools which would get copied anyways

It has nothing to do with move or copy semantics, the key point is release but not managed.

> but would be better to explicitly copy the PlasmaObject instead

No. The general practice is to transferring the ownership by move.

Reply:
> Because the it->second->object can contain a reference to the memory that is still owned by the unique_ptr (which will be freed after this function exits). Might be missing something, not a cpp expert.

No, the pointer pointing to the object is still valid, even if the object being pointed is changed due to move semantics.

Reply:
For my education, is there a practical difference between moving vs. copying the `PlasmaObject` here?

My understanding is that because it is a struct of ints, the default move constructor will do a member-wise move, which will effectively be the same as a copy.

Reply:
> For my education, is there a practical difference between moving vs. copying the PlasmaObject here?

No.

move should be here because:
- It's never guaranteed we won't have non-POD type data member in the future
- move indicates ownership transfer, which is the case here; as I mentioned above, unique_ptr<Object>'s lifecycle is scoped within the function